### PR TITLE
Factory for authority and application

### PR DESCRIPTION
**Pull request by tuler**
Adds a convenience contract that integrates the AuthorityHistoryPairFactory and the CartesiDAppFactory

### CODE REVIEW

File: packages/contracts/contracts/self-hosted/ApplicationFactory.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import {IAuthorityHistoryPairFactory} from "@cartesi/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol";
+import {IConsensus} from "@cartesi/rollups/contracts/consensus/IConsensus.sol";
+import {CartesiDApp} from "@cartesi/rollups/contracts/dapp/CartesiDApp.sol";
+import {CartesiDAppFactory} from "@cartesi/rollups/contracts/dapp/CartesiDAppFactory.sol";
+
+import {IApplicationFactory} from "./IApplicationFactory.sol";
+
+/// @notice This contract implements an IApplicationFactory.
+contract ApplicationFactory is IApplicationFactory {
+    IAuthorityHistoryPairFactory immutable authorityFactory;
```
Comment (guidanoli) [inline packages/contracts/contracts/self-hosted/ApplicationFactory.sol:15]: The name `authorityHistoryPairFactory` would be more accurate.

**Review by guidanoli — COMMENTED** [review]
I would change the name `ApplicationFactory`, because in v2 we've renamed `CartesiDAppFactory` to that. Maybe `SelfHostedApplicationFactory` or something like that?

**Comment by endersonmaia** [comment]
It's failing on my machine, i did a `pnpm clean` and tried to build only `@sunodo/contracts` but it fails.

```shell
❯ pnpm build --filter @sunodo/contracts

> sunodo-monorepo@0.0.0 build /home/endersonmaia/Development/sunodo/sunodo
> turbo run build "--filter" "@sunodo/contracts"

╭───────────────────────────────────────────────────────────────────────╮
│                                                                       │
│                  Update available v1.10.15 ≫ v1.12.5                  │
│    Changelog: https://github.com/vercel/turbo/releases/tag/v1.12.5    │
│           Run "npx @turbo/codemod@latest update" to update            │
│                                                                       │
│        Follow @turborepo for updates: https://x.com/turborepo         │
╰───────────────────────────────────────────────────────────────────────╯
• Packages in scope: @sunodo/contracts
• Running build in 1 packages
• Remote caching disabled
@sunodo/contracts:build: cache miss, executing 739e173884c9a59e
@sunodo/contracts:build: 
@sunodo/contracts:build: > @sunodo/contracts@0.5.0 build /home/endersonmaia/Development/sunodo/sunodo/packages/contracts
@sunodo/contracts:build: > run-s compile tsc export docgen
@sunodo/contracts:build: 
@sunodo/contracts:build: 
@sunodo/contracts:build: > @sunodo/contracts@0.5.0 compile /home/endersonmaia/Development/sunodo/sunodo/packages/contracts
@sunodo/contracts:build: > hardhat compile
@sunodo/contracts:build: 
@sunodo/contracts:build: Nothing to compile
@sunodo/contracts:build: 
@sunodo/contracts:build: > @sunodo/contracts@0.5.0 tsc /home/endersonmaia/Development/sunodo/sunodo/packages/contracts
@sunodo/contracts:build: > tsc
@sunodo/contracts:build: 
@sunodo/contracts:build: test/SelfHostedApplicationFactory.ts(5,1): error TS2582: Cannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
@sunodo/contracts:build: test/SelfHostedApplicationFactory.ts(6,5): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
@sunodo/contracts:build: test/SelfHostedApplicationFactory.ts(41,5): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
@sunodo/contracts:build:  ELIFECYCLE  Command failed with exit code 2.
@sunodo/contracts:build: ERROR: "tsc" exited with 1.
@sunodo/contracts:build:  ELIFECYCLE  Command failed with exit code 1.
@sunodo/contracts:build: ERROR: command finished with error: command (/home/endersonmaia/Development/sunodo/sunodo/packages/contracts) pnpm run build exited (1)
@sunodo/contracts#build: command (/home/endersonmaia/Development/sunodo/sunodo/packages/contracts) pnpm run build exited (1)

 Tasks:    0 successful, 1 total
Cached:    0 cached, 1 total
  Time:    4.624s 
Failed:    @sunodo/contracts#build

 ERROR  run failed: command  exited (1)
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by endersonmaia** [comment]
I had to add @types/jest do packages/contracts.

```shell
cd packages/contracts
pnpm i --save-dev @types/jest
```

**Comment by tuler** [comment]
> I had to add @types/jest do packages/contracts.
> 
> ```shell
> cd packages/contracts
> pnpm i --save-dev @types/jest
> ```

That's strange. `@types/mocha` should be enough.

**Review by guidanoli — APPROVED** [review]
Tested locally and things seem to work.

**Review by endersonmaia — APPROVED** [review]
LGTM :rocket:

**Pull request merged**